modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessExecuteAction.java

```diff
@@ -283,30 +412,89 @@ public PainlessTestScript(Map<String, Object> params) {
-                               ScriptService scriptService) {
+                               ScriptService scriptService, ClusterService clusterService, IndicesService indicesServices) {
             super(settings, NAME, threadPool, transportService, actionFilters, indexNameExpressionResolver, Request::new);
             this.scriptService = scriptService;
+            this.clusterService = clusterService;
+            this.indicesServices = indicesServices;
         }
         @Override
         protected void doExecute(Request request, ActionListener<Response> listener) {
             switch (request.context) {
                 case PAINLESS_TEST:
-                    PainlessTestScript.Factory factory = scriptService.compile(request.script, PainlessTestScript.CONTEXT);
-                    PainlessTestScript painlessTestScript = factory.newInstance(request.script.getParams());
-                    String result = Objects.toString(painlessTestScript.execute());
-                    listener.onResponse(new Response(result));
+                    {
+                        PainlessTestScript.Factory factory = scriptService.compile(request.script, PainlessTestScript.CONTEXT);
+                        PainlessTestScript painlessTestScript = factory.newInstance(request.script.getParams());
+                        String result = Objects.toString(painlessTestScript.execute());
+                        listener.onResponse(new Response(result));
+                    }
+                    break;
+                case FILTER_SCRIPT:
+                    {
+                        indexAndOpenIndexReader(request, (context, leafReaderContext) -> {
+                            FilterScript.Factory factory = scriptService.compile(request.script, FilterScript.CONTEXT);
+                            FilterScript.LeafFactory leafFactory =
+                                factory.newFactory(request.getScript().getParams(), context.lookup());
+                            FilterScript filterScript = leafFactory.newInstance(leafReaderContext);
+                            boolean result = filterScript.execute();
+                            listener.onResponse(new Response(result));
+                        });
+                    }
+                    break;
+                case SEARCH_SCRIPT:
+                    {
+                        indexAndOpenIndexReader(request, (context, leafReaderContext) -> {
+                            SearchScript.Factory factory = scriptService.compile(request.script, SearchScript.CONTEXT);
+                            SearchScript.LeafFactory leafFactory =
+                                factory.newFactory(request.getScript().getParams(), context.lookup());
+                            SearchScript searchScript = leafFactory.newInstance(leafReaderContext);
+                            Object result = searchScript.run();
```

Comment:
This is another reason I would like script fields and sort scripts separated out. Actual scoring scripts should call the method expecting a double here.